File: ImasLiveDB-Android/app/src/main/kotlin/com/fugaif/imaslivedb/i18n/DisplayLocale.kt
package com.fugaif.imaslivedb.i18n

import android.content.Context
import com.fugaif.imaslivedb.i18n.generated.L10n
import java.util.IllformedLocaleException
import java.util.Locale

/**
 * 画面に実際に出ている言語 (iOS `DisplayLocale` と対)。
 *
 * 呼び出し時点の [Context] でカタログの予約キー `i18n.language_tag` を解決して得る。アプリが対応しない
 * 言語の端末では UI がフォールバック先 (今は ja) の言語で出るので、端末の言語 (Locale.getDefault()) ではなく
 * 実際に引けた表の値を使う。こうすると日付の書式の言語が UI とずれない。アプリ全体でキャッシュした
 * Resources も使わない (アプリ別の言語を切り替えると古いまま残る)。
 *
 * @property languageTag 表示言語のコード ("ja", "ko")。
 * @property formattingLocale 日付・数値の書式に使うロケール。言語と文字体系は [languageTag]、地域などは端末のもの。
 */
data class DisplayLocale(val languageTag: String, val formattingLocale: Locale) {
    companion object {
        fun of(context: Context): DisplayLocale {
            val tag = L10n.I18n.languageTag.resolve(context)
            val language = Locale.forLanguageTag(tag)
            val device = context.resources.configuration.locales[0]
            val formatting = try {
                // 言語と文字体系を差し替える (端末の文字体系は残さない。"sr-Latn" の端末で ja にしたとき
                // "ja-Latn" にならないように。iOS の DisplayLocale と同じ結果)。setLanguage は "zh-Hans" を受けない
                Locale.Builder().setLocale(device).setLanguage(language.language).setScript(language.script).build()
            } catch (e: IllformedLocaleException) {
                // 端末のロケールが BCP 47 として組めない (古い variant など) ときは言語だけで書式を決める。
                language
            }
            return DisplayLocale(tag, formatting)
        }
    }
}

File: ImasLiveDB/Shared/L10n/DisplayLocale.swift
import Foundation

/// 画面に実際に出ている言語。`Locale.current` ではなく、カタログの `i18n.language_tag` を解決して得る。
///
/// 未対応言語の端末でフォールバック中でも、実際に引かれた表の言語になる。
/// 出荷ゲート (XCSTRINGS_LANGUAGES_TO_COMPILE) で外した言語も自動で反映される。
/// Android の `DisplayLocale` と同じ規則。
struct DisplayLocale: Sendable, Equatable {
    /// "ja" / "ko"。カタログの i18n.language_tag の値そのもの。
    let languageTag: String
    /// 日付・数値の書式に使うロケール。言語と文字体系は languageTag、地域などは端末設定。
    let formattingLocale: Locale

    /// 呼ぶたびに表を引く (キャッシュしない。LocalizedStringResource と同じく作った時点の言語で固まるため)。
    static var current: DisplayLocale {
        let tag = String(localized: L10n.I18n.languageTag)
        let language = Locale.Language.Components(identifier: tag)
        var components = Locale.Components(locale: .current)
        // languageComponents ごと差し替えると地域 (region) まで消える。言語と文字体系だけ差し替える
        // (端末の文字体系は残さない。"sr-Latn" の端末で ja にしたとき "ja-Latn" にならないように)
        components.languageComponents.languageCode = language.languageCode
        components.languageComponents.script = language.script
        return DisplayLocale(languageTag: tag, formattingLocale: Locale(components: components))
    }
}
